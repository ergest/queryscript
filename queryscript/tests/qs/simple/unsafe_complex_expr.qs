import * from inference;

-- This tests to make sure that _after_ resolving schemas, but _before_ resolving unsafe expressions, the compiler has
-- enough oxygen to infer types (e.g. for the complex_expr variable in the case below).
--
-- I kind of caught this test case like lightning in a bottle. It is a bit hard / hacky to get it to reproduce
-- like this. But the basic intuition is that if there's enough complexity in the expression, we'll have some
-- work to do between figuring out the type of users (via schema inference) and figuring out the type of complex_expr
-- that we'll fail to compile the unsafe query below if we don't let the handles run in between.
let complex_expr  = SELECT
    id,
    org_id,
    name,
    active,
    id > 0 or active,
    (SELECT id FROM users i WHERE i.id = o.id AND i.org_id = o.org_id AND i.active AND o.active) max_user
  FROM users o
  WHERE active AND id < 1 AND id > 5
;

unsafe select * from complex_expr;
