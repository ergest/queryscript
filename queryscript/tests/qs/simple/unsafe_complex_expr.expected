{
    "compile_errors": [],
    "decls": {
        "let complex_expr": [{
        	id Int64,
        	org_id Int64,
        	name Utf8,
        	active Boolean,
        	id > 0 OR active Boolean,
        	max_user Int64,
        }],
        "let events": External<[{
        	description Utf8,
        	ts Utf8,
        	user_id Int64,
        }]>,
        "let users": External<[{
        	active Boolean,
        	id Int64,
        	name Utf8,
        	org_id Int64,
        }]>,
    },
    "queries": [
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "name",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "active",
                                type_: Atom(
                                    Boolean,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "id > 0 OR active",
                                type_: Atom(
                                    Boolean,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "max_user",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | org_id | name | active | id > 0 OR active | max_user |",
            },
        ),
    ],
}